[Fix] OpenRouter Streaming Usage Cost

dhruvyad opened the pull request:
## Title

Fix usage cost calculation for streaming models via OpenRouter.

## Relevant issues

Extension of #13653

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [ ] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

<img width="1259" height="865" alt="image" src="https://github.com/user-attachments/assets/26a33362-58cb-4456-aa9c-cf173d9270cb" />

## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->


🐛 Bug Fix
✅ Test

## Changes

Fix usage cost tracking for streaming, which is currently broken for OpenRouter models.

dhruvyad committed:
fix openrouter streaming usage

dhruvyad commented:
> Is this PR still active? I'd love to have this feature

I'm using a forked version cause this PR didn't get reviewed. Happy to rebase and address comments if there's interest in merging this.

mparsam commented:
As of today really look forward to have this fix 
the support for non-streaming is done [here](https://github.com/BerriAI/litellm/pull/15448)
Streaming is the real need here.  only me, I know at least two teams completely blocked by this problem.
@dhruvyad @krrishdholakia 

or if anyone has trick to solve this I will be thankful

dhruvyad committed:
propagate streaming usage.cost to _hidden_params for cost tracking

Without this, the provider-reported cost (e.g. from OpenRouter) was
available on usage.cost but never reached litellm's cost calculator,
which reads from _hidden_params["additional_headers"]. Also cleans up
setattr usage in tests since Usage already has a cost field.

dhruvyad committed:
apply black formatting

dhruvyad committed:
address review feedback: remove provider-specific guard, fix double-append, DRY cost propagation

- Remove hardcoded openrouter check from core handler; the general
  return-instead-of-raise in return_processed_chunk_logic already
  handles post-finish-reason usage chunks for any provider.
- Remove self.chunks.append in return_processed_chunk_logic to avoid
  double-appending (caller already appends).
- Build proper Usage in calculate_total_usage instead of returning
  raw chunk early — preserves cost while ensuring token fields are
  always reconstructed.
- Extract cost propagation into _propagate_usage_cost_to_hidden_params
  static method, used by both sync and async paths.
- Test calls the real method instead of copy-pasting production logic.

arbv commented:
not stale

blakeaa827 commented:
Reviewed and tested this locally (checked out this branch, Python 3.9) — it correctly fixes the core of #11626 (which is still open, auto-closed as stale).

**Mechanism looks right:** `_propagate_usage_cost_to_hidden_params` writes `usage.cost` into `_hidden_params["additional_headers"]["llm_provider-x-litellm-response-cost"]` — the same key the non-streaming path (`OpenrouterConfig.transform_response`) uses, and which `get_response_cost_from_hidden_params` short-circuits on before the static cost map. So streamed OpenRouter calls now report the real provider cost instead of falling back to the map (which is `$0` for models that aren't in it).

**Tests pass locally:**
```
tests/test_litellm/litellm_core_utils/test_streaming_handler.py
tests/test_litellm/litellm_core_utils/test_streaming_chunk_builder_utils.py
tests/test_litellm/llms/openrouter/
→ 166 passed
```

**One gap:** this threads `cost` but not `cost_details` / `is_byok`, and #11626's title explicitly names `is_byok`. It's a small addition right after `returned_usage_chunk.cost = latest_usage_chunk.cost` in `calculate_total_usage`:

```python
    # Preserve OpenRouter-style provider usage metadata (issue #11626).
    if getattr(latest_usage_chunk, "cost_details", None) is not None:
        returned_usage_chunk.cost_details = latest_usage_chunk.cost_details
    if getattr(latest_usage_chunk, "is_byok", None) is not None:
        returned_usage_chunk.is_byok = latest_usage_chunk.is_byok
```

I have this plus a test passing locally (167 total) and am happy to send it as a follow-up once this merges, or you're welcome to fold it in here. Either way, solid fix — would be great to see it land. It resolves #11626.

krrish-berri-2 commented:
+ @Mateo Wang ***@***.***> for review

On Fri, Jul 3, 2026 at 8:16 PM blakeaa827 ***@***.***> wrote:

> *blakeaa827* left a comment (BerriAI/litellm#16162)
> <https://github.com/BerriAI/litellm/pull/16162#issuecomment-4880461791>
>
> Reviewed and tested this locally (checked out this branch, Python 3.9) —
> it correctly fixes the core of #11626
> <https://github.com/BerriAI/litellm/issues/11626> (which is still open,
> auto-closed as stale).
>
> *Mechanism looks right:* _propagate_usage_cost_to_hidden_params writes
> usage.cost into
> _hidden_params["additional_headers"]["llm_provider-x-litellm-response-cost"]
> — the same key the non-streaming path (OpenrouterConfig.transform_response)
> uses, and which get_response_cost_from_hidden_params short-circuits on
> before the static cost map. So streamed OpenRouter calls now report the
> real provider cost instead of falling back to the map (which is $0 for
> models that aren't in it).
>
> *Tests pass locally:*
>
> tests/test_litellm/litellm_core_utils/test_streaming_handler.py
> tests/test_litellm/litellm_core_utils/test_streaming_chunk_builder_utils.py
> tests/test_litellm/llms/openrouter/
> → 166 passed
>
> *One gap:* this threads cost but not cost_details / is_byok, and #11626
> <https://github.com/BerriAI/litellm/issues/11626>'s title explicitly
> names is_byok. It's a small addition right after returned_usage_chunk.cost
> = latest_usage_chunk.cost in calculate_total_usage:
>
>     # Preserve OpenRouter-style provider usage metadata (issue #11626).
>     if getattr(latest_usage_chunk, "cost_details", None) is not None:
>         returned_usage_chunk.cost_details = latest_usage_chunk.cost_details
>     if getattr(latest_usage_chunk, "is_byok", None) is not None:
>         returned_usage_chunk.is_byok = latest_usage_chunk.is_byok
>
> I have this plus a test passing locally (167 total) and am happy to send
> it as a follow-up once this merges, or you're welcome to fold it in here.
> Either way, solid fix — would be great to see it land. It resolves #11626
> <https://github.com/BerriAI/litellm/issues/11626>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BerriAI/litellm/pull/16162?email_source=notifications&email_token=CARFWGF6IPK5DDECQN76LJD5DBZH5A5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIOBYGA2DMMJXHEY2M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4880461791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/CARFWGEVJBFKGCGGTA6KEGT5DBZH5AVCNFSNUABFKJSXA33TNF2G64TZHM3DOMJSGY4TKMBVHNEXG43VMU5TGNJXHA2DCMRWGAY2C5QC>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
*Krrish Dholakia | *CEO
Book a meeting with me
<https://calendly.com/d/cx9p-5yf-2nm/litellm-introductions>
LinkedIn <https://www.linkedin.com/in/krish-d/> | (770) 878 - 3106
***@***.***

P.S. See how LiteLLM helps LLM Platform teams move fast and stay in control
<https://www.litellm.ai/#features>

mateo-berri commented:
I'll make a copy branch for review. I see a bunch of folks testing via unit tests. @dhruvyad do you mind uploading screenshots/video or an ordered list of command+command outputs ran showing it works e2e?
